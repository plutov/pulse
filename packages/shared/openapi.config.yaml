withInterfaces: true
withSeparateModelsAndApi: true
apiPackage: api
modelPackage: models
enumPropertyNaming: camelCase
apiDocs: false
modelDocs: false
apiTests: false
modelTests: false
npmName: "@pulse/apigen"
npmVersion: "0.0.0"

