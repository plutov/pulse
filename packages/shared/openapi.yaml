openapi: 3.0.0
info:
  title: Pulse API
  description: ""
  version: 0.0.0
paths:
  /auth/login:
    post:
      tags:
        - auth
      summary: User Login
      operationId: login
      requestBody:
        $ref: "#/components/requestBodies/LoginBody"
      responses:
        "200":
          description: Login successful
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/LoginResponse"
        "401":
          $ref: "#/components/responses/UnauthorizedError"
        "500":
          $ref: "#/components/responses/InternalServerError"
  /monitors:
    get:
      tags:
        - monitor
      security:
        - BearerAuth: []
      summary: List Monitors
      operationId: listMonitors
      responses:
        "200":
          description: ok
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: "#/components/schemas/Monitor"
        "401":
          $ref: "#/components/responses/UnauthorizedError"
        "500":
          $ref: "#/components/responses/InternalServerError"
    post:
      tags:
        - monitor
      security:
        - BearerAuth: []
      summary: Create Monitor
      operationId: createMonitor
      requestBody:
        $ref: "#/components/requestBodies/CreateMonitorBody"
      responses:
        "201":
          description: ok
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Monitor"
        "400":
          $ref: "#/components/responses/BadRequestError"
        "401":
          $ref: "#/components/responses/UnauthorizedError"
        "500":
          $ref: "#/components/responses/InternalServerError"
  /monitors/{id}:
    get:
      tags:
        - monitor
      security:
        - BearerAuth: []
      summary: Get Monitor
      operationId: getMonitor
      parameters:
        - $ref: "#/components/parameters/MonitorId"
      responses:
        "200":
          description: ok
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Monitor"
        "400":
          $ref: "#/components/responses/BadRequestError"
        "401":
          $ref: "#/components/responses/UnauthorizedError"
        "404":
          $ref: "#/components/responses/NotFoundError"
        "500":
          $ref: "#/components/responses/InternalServerError"
    delete:
      tags:
        - monitor
      security:
        - BearerAuth: []
      summary: Delete Monitor
      operationId: deleteMonitor
      parameters:
        - $ref: "#/components/parameters/MonitorId"
      responses:
        "204":
          description: Monitor deleted successfully
        "400":
          $ref: "#/components/responses/BadRequestError"
        "401":
          $ref: "#/components/responses/UnauthorizedError"
        "404":
          $ref: "#/components/responses/NotFoundError"
        "500":
          $ref: "#/components/responses/InternalServerError"
  /runs:
    get:
      tags:
        - monitor
      security:
        - BearerAuth: []
      summary: Get Monitor Runs
      operationId: listMonitorRuns
      parameters:
        - $ref: "#/components/parameters/MonitorIdQuery"
        - $ref: "#/components/parameters/Size"
        - $ref: "#/components/parameters/Offset"
      responses:
        "200":
          description: ok
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/MonitorRunsList"
        "400":
          $ref: "#/components/responses/BadRequestError"
        "401":
          $ref: "#/components/responses/UnauthorizedError"
        "404":
          $ref: "#/components/responses/NotFoundError"
        "500":
          $ref: "#/components/responses/InternalServerError"
  /charts/timeSeries:
    post:
      tags:
        - charts
      security:
        - BearerAuth: []
      summary: Get Time Series Chart Data
      operationId: getTimeSeriesChartData
      requestBody:
        $ref: "#/components/requestBodies/TimeSeriesChartBody"
      responses:
        "200":
          description: ok
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/TimeSeriesChartData"
        "400":
          $ref: "#/components/responses/BadRequestError"
        "401":
          $ref: "#/components/responses/UnauthorizedError"
        "500":
          $ref: "#/components/responses/InternalServerError"
components:
  securitySchemes:
    BearerAuth:
      type: http
      scheme: bearer
      bearerFormat: JWT
  schemas:
    ValidationMessage:
      required:
        - message
        - type
      type: object
      properties:
        message:
          type: string
        type:
          type: string
        path:
          type: array
          items:
            type: string
    ErrorResponse:
      required:
        - statusCode
        - message
        - validationMessages
      type: object
      properties:
        statusCode:
          type: integer
        message:
          type: string
        validationMessages:
          type: array
          items:
            $ref: "#/components/schemas/ValidationMessage"
    WithId:
      required:
        - id
      type: object
      properties:
        id:
          type: string
          format: uuid
    WithName:
      required:
        - name
      type: object
      properties:
        name:
          type: string
    MonitorType:
      type: string
      enum:
        - http
    WithMonitorType:
      required:
        - monitorType
      type: object
      properties:
        monitorType:
          $ref: "#/components/schemas/MonitorType"
    WithCreatedAt:
      required:
        - createdAt
      type: object
      properties:
        createdAt:
          type: string
          format: date-time
    WithUpdatedAt:
      required:
        - updatedAt
      type: object
      properties:
        updatedAt:
          type: string
          format: date-time
    User:
      allOf:
        - $ref: "#/components/schemas/WithId"
        - $ref: "#/components/schemas/WithUsername"
    WithAuthor:
      required:
        - author
      type: object
      properties:
        author:
          $ref: "#/components/schemas/User"
    MonitorConfig:
      oneOf:
        - $ref: "#/components/schemas/HttpConfig"
    HttpConfig:
      required:
        - url
        - method
      type: object
      properties:
        url:
          type: string
        method:
          type: string
          enum:
            - GET
            - POST
            - PUT
            - DELETE
            - HEAD
            - OPTIONS
            - PATCH
    WithConfig:
      type: object
      required:
        - config
      properties:
        config:
          $ref: "#/components/schemas/MonitorConfig"
    WithSchedule:
      required:
        - schedule
      type: object
      properties:
        schedule:
          type: string
          description: Cron-like schedule string
          example: "* * * * *"
    MonitorStatus:
      type: string
      enum:
        - active
        - paused
    WithStatus:
      required:
        - status
      type: object
      properties:
        status:
          $ref: "#/components/schemas/MonitorStatus"
    Monitor:
      allOf:
        - $ref: "#/components/schemas/WithId"
        - $ref: "#/components/schemas/WithName"
        - $ref: "#/components/schemas/WithMonitorType"
        - $ref: "#/components/schemas/WithCreatedAt"
        - $ref: "#/components/schemas/WithUpdatedAt"
        - $ref: "#/components/schemas/WithAuthor"
        - $ref: "#/components/schemas/WithSchedule"
        - $ref: "#/components/schemas/WithStatus"
        - $ref: "#/components/schemas/WithConfig"
    MonitorMinimal:
      allOf:
        - $ref: "#/components/schemas/WithId"
        - $ref: "#/components/schemas/WithName"
    WithMonitorMinimal:
      required:
        - monitor
      type: object
      properties:
        monitor:
          $ref: "#/components/schemas/MonitorMinimal"
    CreateMonitorPayload:
      allOf:
        - $ref: "#/components/schemas/WithName"
        - $ref: "#/components/schemas/WithMonitorType"
        - $ref: "#/components/schemas/WithSchedule"
        - $ref: "#/components/schemas/WithStatus"
        - $ref: "#/components/schemas/WithConfig"
    WithUsername:
      required:
        - username
      type: object
      properties:
        username:
          type: string
    WithPassword:
      required:
        - password
      type: object
      properties:
        password:
          type: string
    LoginPayload:
      allOf:
        - $ref: "#/components/schemas/WithUsername"
        - $ref: "#/components/schemas/WithPassword"
    WithToken:
      required:
        - token
      type: object
      properties:
        token:
          type: string
    LoginResponse:
      allOf:
        - $ref: "#/components/schemas/WithToken"
        - $ref: "#/components/schemas/WithUser"
    WithUser:
      required:
        - user
      type: object
      properties:
        user:
          $ref: "#/components/schemas/User"
    MonitorsList:
      type: object
      required:
        - total
        - rows
      properties:
        total:
          type: integer
        rows:
          type: array
          items:
            $ref: "#/components/schemas/Monitor"
    MonitorRunsList:
      type: object
      required:
        - total
        - rows
      properties:
        total:
          type: integer
        rows:
          type: array
          items:
            $ref: "#/components/schemas/MonitorRun"
    WithRunStatus:
      required:
        - status
      type: object
      properties:
        status:
          $ref: "#/components/schemas/RunStatus"
    RunStatus:
      type: string
      enum:
        - success
        - failure
        - timeout
    WithDurationMs:
      required:
        - durationMs
      type: object
      properties:
        durationMs:
          type: integer
    WithMonitorRunDetails:
      required:
        - details
      type: object
      properties:
        details:
          $ref: "#/components/schemas/MonitorRunDetails"
    MonitorRunDetails:
      oneOf:
        - $ref: "#/components/schemas/HttpRunDetails"
    HttpRunDetails:
      required:
        - statusCode
      properties:
        statusCode:
          type: integer
    MonitorRun:
      allOf:
        - $ref: "#/components/schemas/WithId"
        - $ref: "#/components/schemas/WithCreatedAt"
        - $ref: "#/components/schemas/WithMonitorMinimal"
        - $ref: "#/components/schemas/WithRunStatus"
        - $ref: "#/components/schemas/WithDurationMs"
        - $ref: "#/components/schemas/WithMonitorRunDetails"
    ChartInterval:
      type: string
      enum:
        - 5m
        - 1h
        - 1d
        - 1w
        - 30d
    TimeSeriesChartPayload:
      required:
        - monitorIds
        - timeRange
        - interval
      type: object
      properties:
        monitorIds:
          type: array
          items:
            type: string
            format: uuid
        start:
          type: string
          format: date-time
        end:
          type: string
          format: date-time
        interval:
          $ref: "#/components/schemas/ChartInterval"
        timeRange:
          $ref: "#/components/schemas/ChartInterval"
    TimeSeriesDataPoint:
      allOf:
        - $ref: "#/components/schemas/WithCreatedAt"
        - $ref: "#/components/schemas/WithRunStatus"
        - $ref: "#/components/schemas/WithDurationMs"
    WithDataPoints:
      required:
        - dataPoints
      type: object
      properties:
        dataPoints:
          type: array
          items:
            $ref: "#/components/schemas/TimeSeriesDataPoint"
    TimeSeriesMonitorData:
      allOf:
        - $ref: "#/components/schemas/WithMonitorMinimal"
        - $ref: "#/components/schemas/WithDataPoints"
    TimeSeriesChartData:
      required:
        - data
      type: object
      properties:
        data:
          type: array
          items:
            $ref: "#/components/schemas/TimeSeriesMonitorData"
  requestBodies:
    LoginBody:
      required: true
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/LoginPayload"
    CreateMonitorBody:
      required: true
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/CreateMonitorPayload"
    TimeSeriesChartBody:
      required: true
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/TimeSeriesChartPayload"
  responses:
    BadRequestError:
      description: "400"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/ErrorResponse"
    UnauthorizedError:
      description: "401"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/ErrorResponse"
    NotFoundError:
      description: "404"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/ErrorResponse"
    InternalServerError:
      description: "500"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/ErrorResponse"
  parameters:
    MonitorId:
      name: id
      in: path
      required: true
      schema:
        type: string
        format: uuid
    MonitorIdQuery:
      name: monitorId
      in: query
      required: false
      schema:
        type: string
        format: uuid
    Size:
      name: size
      in: query
      required: false
      schema:
        type: integer
        default: 50
        minimum: 1
        maximum: 1000
    Offset:
      name: offset
      in: query
      required: false
      schema:
        type: integer
        default: 0
        minimum: 0
