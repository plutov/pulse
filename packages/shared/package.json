{
  "name": "@pulse/shared",
  "version": "1.0.0",
  "private": true,
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist",
    "generate": "rm -rf ./src/apigen && openapi-generator-cli generate -i ./openapi.yaml -g typescript-axios -o ./src/apigen -c openapi.config.yaml"
  },
  "devDependencies": {
    "@openapitools/openapi-generator-cli": "^2.23.1",
    "typescript": "^5.9.2"
  },
  "files": [
    "dist"
  ]
}
