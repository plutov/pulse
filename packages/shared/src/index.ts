// Re-export all generated types and APIs
export * from './apigen/src';