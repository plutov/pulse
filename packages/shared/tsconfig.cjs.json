{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "noEmit": false,
    "target": "ESNext",
    "module": "CommonJS",
    "outDir": "dist/cjs",
    "declaration": true,
    "declarationDir": "dist/types"
  }
}
