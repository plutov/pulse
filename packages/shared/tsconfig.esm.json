{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "noEmit": false,
    "module": "ESNext",
    "outDir": "dist/esm",
    "declaration": true,
    "declarationDir": "dist/types",
    "moduleResolution": "node"
  }
}
