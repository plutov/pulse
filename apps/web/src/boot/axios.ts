import { defineBoot } from "#q-app/wrappers";
import axios, { type AxiosInstance } from "axios";
import { AuthApi, MonitorApi, Configuration } from "@pulse/shared";
import { useAuthStore } from "stores/auth";
import { unref } from "vue";

declare module "vue" {
  interface ComponentCustomProperties {
    $axios: AxiosInstance;
    $api: AxiosInstance;
  }
}

const createConfig = () => {
  return new Configuration({
    basePath: process.env.API_ADDR || "http://localhost:3000",
    accessToken: () => unref(useAuthStore().token),
  });
};

export const getAuthApi = () => new AuthApi(createConfig());
export const getMonitorApi = () => new MonitorApi(createConfig());

export const authApi = new AuthApi(createConfig());
export const monitorApi = new MonitorApi(createConfig());

export default defineBoot(({ app }) => {
  const authStore = useAuthStore();
  authStore.initializeAuth();

  app.config.globalProperties.$authStore = authStore;
  app.config.globalProperties.$axios = axios;
});
