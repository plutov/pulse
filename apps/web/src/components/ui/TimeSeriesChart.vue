<template>
  <div class="time-series-chart">
    <q-card>
      <q-card-section>
        <div class="row items-center justify-between">
          <div>
            <div class="text-h6">{{ title }}</div>
            <div class="text-subtitle2 text-grey">
              Showing results over {{ timeRange }} ({{ interval }} intervals)
            </div>
          </div>
          <q-btn
            icon="refresh"
            round
            flat
            :loading="loading"
            @click="fetchChartData"
            :disable="!monitorIds.length"
          >
            <q-tooltip>Refresh chart</q-tooltip>
          </q-btn>
        </div>
      </q-card-section>
      <q-card-section>
        <div v-if="loading" class="row justify-center q-pa-md">
          <q-spinner-dots size="lg" />
        </div>
        <div v-else-if="error" class="text-negative q-pa-md">
          {{ error }}
        </div>
        <canvas ref="chartCanvas" id="chartCanvas"></canvas>
      </q-card-section>
    </q-card>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, watch, nextTick } from "vue";
import { chartsApi } from "src/boot/axios";
import type {
  TimeSeriesChartPayload,
  TimeSeriesChartData,
  ChartInterval,
} from "@pulse/shared";
import {
  Chart,
  CategoryScale,
  LinearScale,
  LineElement,
  PointElement,
  Title,
  Tooltip,
  Legend,
  TimeScale,
  LineController,
} from "chart.js";
import "chartjs-adapter-date-fns";
import { chartColors } from "./colors";
import { DateTimeFormat } from "src/api/time";

Chart.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  LineController,
  Title,
  Tooltip,
  Legend,
  TimeScale,
);

interface Props {
  title: string;
  monitorIds: string[];
  timeRange: ChartInterval;
  interval: ChartInterval;
}

const props = withDefaults(defineProps<Props>(), {});

const chartCanvas = ref<HTMLCanvasElement>();
const loading = ref(false);
const error = ref("");
let chartInstance: Chart | null = null;

const statusColors: Record<string, string> = {
  success: chartColors.positive,
  failure: chartColors.negative,
  timeout: chartColors.warning,
};

const lineColors = [
  chartColors.primary,
  chartColors.warning,
  chartColors.info,
  chartColors.secondary,
  chartColors.accent,
];

const fetchChartData = async () => {
  if (!props.monitorIds.length) return;

  loading.value = true;
  error.value = "";

  try {
    const payload: TimeSeriesChartPayload = {
      monitorIds: props.monitorIds,
      timeRange: props.timeRange,
      interval: props.interval,
    };

    const response = await chartsApi.getTimeSeriesChartData(payload);
    const chartData = response.data;

    if (chartCanvas.value) {
      createChart(chartData);
    }
  } catch {
    error.value = "Failed to load chart data";
  } finally {
    loading.value = false;
  }
};

const createChart = (data: TimeSeriesChartData) => {
  if (!chartCanvas.value) return;

  if (chartInstance) {
    chartInstance.destroy();
  }

  const ctx = chartCanvas.value.getContext("2d");
  if (!ctx) return;

  // Generate datasets for each monitor
  const datasets = data.data.map((monitor, index) => {
    const color = lineColors[index % lineColors.length];

    return {
      label: monitor.monitor.name,
      data: monitor.dataPoints.map((point) => ({
        x: new Date(point.createdAt).getTime(),
        y: point.durationMs,
      })),
      borderColor: color,
      backgroundColor: color + "20",
      fill: false,
      tension: 0.3,
      pointBackgroundColor: monitor.dataPoints.map((point) => {
        const statusColor = statusColors[point.status as string];
        return statusColor || color;
      }),
      pointBorderColor: "",
      pointRadius: 2,
      pointHoverRadius: 4,
    };
  });

  chartInstance = new Chart(ctx, {
    type: "line",
    data: { datasets },
    options: {
      responsive: true,
      maintainAspectRatio: false,
      layout: {
        padding: {
          top: 10,
          bottom: 10,
        },
      },
      interaction: {
        intersect: false,
        mode: "index",
      },
      scales: {
        x: {
          type: "time",
          title: {
            display: true,
            text: "Time",
          },
        },
        y: {
          title: {
            display: true,
            text: "Duration (ms)",
          },
          beginAtZero: true,
        },
      },
      plugins: {
        tooltip: {
          callbacks: {
            label: (context) => {
              const monitorData = data.data[context.datasetIndex];
              const dataPoint = monitorData?.dataPoints[context.dataIndex];
              if (!dataPoint) {
                return `${context.dataset.label}: ${context.parsed.y}ms`;
              }
              return `${context.dataset.label}: ${context.parsed.y}ms (${dataPoint.status})`;
            },
          },
        },
        legend: {
          position: "top",
        },
      },
    },
  });
};

onMounted(async () => {
  await nextTick();
  await fetchChartData();
});

watch(
  () => [props.monitorIds, props.timeRange, props.interval],
  () => {
    void fetchChartData();
  },
  { deep: true },
);
</script>

<style scoped>
.time-series-chart {
  width: 100%;
  height: 400px;
}

.time-series-chart .q-card-section:last-child {
  height: 300px;
  position: relative;
}

canvas {
  width: 100% !important;
  height: 300px !important;
}
</style>
