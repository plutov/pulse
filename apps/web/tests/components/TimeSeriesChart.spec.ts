import { mount } from "@vue/test-utils";
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { createApp } from "vue";
import { createPinia, setActivePinia } from "pinia";
import TimeSeriesChart from "../../src/components/ui/TimeSeriesChart.vue";
import type { TimeSeriesChartData } from "@pulse/shared";
import { ChartInterval, RunStatus } from "@pulse/shared";

// Mock Chart.js
const mockChartInstance = {
  destroy: vi.fn(),
};

vi.mock("chart.js", () => {
  const Chart = Object.assign(
    vi.fn().mockImplementation(() => mockChartInstance),
    {
      register: vi.fn(),
    },
  );
  return {
    Chart,
    CategoryScale: vi.fn(),
    LinearScale: vi.fn(),
    LineElement: vi.fn(),
    PointElement: vi.fn(),
    Title: vi.fn(),
    Tooltip: vi.fn(),
    Legend: vi.fn(),
    TimeScale: vi.fn(),
    LineController: vi.fn(),
  };
});

// Mock chartjs-adapter-date-fns
vi.mock("chartjs-adapter-date-fns", () => ({}));

// Mock the charts API
vi.mock("src/boot/axios", () => ({
  chartsApi: {
    getTimeSeriesChartData: vi.fn(),
  },
}));

import { chartsApi } from "src/boot/axios";

describe("TimeSeriesChart.vue", () => {
  let wrapper: ReturnType<typeof mount>;

  const mockChartData: TimeSeriesChartData = {
    data: [
      {
        monitor: {
          id: "1",
          name: "Test Monitor 1",
        },
        dataPoints: [
          {
            createdAt: "2023-01-01T00:00:00.000Z",
            durationMs: 100,
            status: RunStatus.success,
          },
          {
            createdAt: "2023-01-01T01:00:00.000Z",
            durationMs: 150,
            status: RunStatus.failure,
          },
        ],
      },
      {
        monitor: {
          id: "2",
          name: "Test Monitor 2",
        },
        dataPoints: [
          {
            createdAt: "2023-01-01T00:00:00.000Z",
            durationMs: 200,
            status: RunStatus.timeout,
          },
        ],
      },
    ],
  };

  const defaultProps = {
    title: "Test Chart",
    monitorIds: ["1", "2"],
    timeRange: ChartInterval._1d,
    interval: ChartInterval._1h,
  };

  beforeEach(() => {
    setActivePinia(createPinia());

    vi.mocked(chartsApi.getTimeSeriesChartData).mockResolvedValue({
      data: mockChartData,
      status: 200,
      statusText: "OK",
      headers: {},
      config: {} as any,
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
    if (wrapper) {
      wrapper.unmount();
    }
  });

  it("renders with default props", async () => {
    wrapper = mount(TimeSeriesChart, {
      props: defaultProps,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.text()).toContain("Test Chart");
    expect(wrapper.text()).toContain("Showing results over 1d (1h intervals)");
  });

  it("renders chart when data is loaded", async () => {
    wrapper = mount(TimeSeriesChart, {
      props: defaultProps,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find("canvas").exists()).toBe(true);
    expect(wrapper.findComponent({ name: "q-spinner-dots" }).exists()).toBe(
      false,
    );
    expect(wrapper.text()).not.toContain("Failed to load chart data");
  });

  it("refresh button is enabled when monitorIds are provided", async () => {
    wrapper = mount(TimeSeriesChart, {
      props: defaultProps,
    });

    await wrapper.vm.$nextTick();

    const refreshButton = wrapper.find("q-btn");
    expect(refreshButton.attributes("disable")).not.toBe("true");
  });

  it("refresh button is disabled when no monitorIds are provided", async () => {
    wrapper = mount(TimeSeriesChart, {
      props: {
        ...defaultProps,
        monitorIds: [],
      },
    });

    await wrapper.vm.$nextTick();

    const refreshButton = wrapper.find("q-btn");
    expect(refreshButton.attributes("disable")).toBeDefined();
  });

  it("calls fetchChartData when refresh button is clicked", async () => {
    wrapper = mount(TimeSeriesChart, {
      props: defaultProps,
    });

    await wrapper.vm.$nextTick();

    const refreshButton = wrapper.find("q-btn");
    await refreshButton.trigger("click");

    expect(vi.mocked(chartsApi.getTimeSeriesChartData)).toHaveBeenCalledTimes(
      2,
    ); // Once on mount, once on click
    expect(
      vi.mocked(chartsApi.getTimeSeriesChartData),
    ).toHaveBeenLastCalledWith({
      monitorIds: ["1", "2"],
      timeRange: ChartInterval._1d,
      interval: ChartInterval._1h,
    });
  });

  it("does not fetch data when monitorIds is empty", async () => {
    vi.mocked(chartsApi.getTimeSeriesChartData).mockClear();

    wrapper = mount(TimeSeriesChart, {
      props: {
        ...defaultProps,
        monitorIds: [],
      },
    });

    await wrapper.vm.$nextTick();

    expect(vi.mocked(chartsApi.getTimeSeriesChartData)).not.toHaveBeenCalled();
  });

  it("displays correct title and subtitle", async () => {
    wrapper = mount(TimeSeriesChart, {
      props: {
        ...defaultProps,
        title: "Custom Title",
        timeRange: ChartInterval._30d,
        interval: ChartInterval._5m,
      },
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.text()).toContain("Custom Title");
    expect(wrapper.text()).toContain("Showing results over 30d (5m intervals)");
  });

  it("shows tooltip on refresh button", async () => {
    wrapper = mount(TimeSeriesChart, {
      props: defaultProps,
    });

    await wrapper.vm.$nextTick();

    const tooltip = wrapper.find("q-tooltip");
    expect(tooltip.text()).toBe("Refresh chart");
  });
});

