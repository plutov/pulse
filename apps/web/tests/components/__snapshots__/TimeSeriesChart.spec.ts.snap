// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimeSeriesChart.vue > renders with default props 1`] = `
"<div class="time-series-chart">
  <q-card>
    <q-card-section>
      <div class="row items-center justify-between">
        <div>
          <div class="text-h6">Test Chart</div>
          <div class="text-subtitle2 text-grey"> Showing results over 1d (1h intervals) </div>
        </div>
        <q-btn icon="refresh" round="" flat="" loading="false" disable="false">
          <q-tooltip>Refresh chart</q-tooltip>
        </q-btn>
      </div>
    </q-card-section>
    <q-card-section>
      <!--v-if--><canvas id="chartCanvas"></canvas>
    </q-card-section>
  </q-card>
</div>"
`;
